### PR TITLE
Add api, commands, and binding support for mTLS certificates

### PR DESCRIPTION
**Pull request by matthewdavidrodgers**
What this PR solves / how to test:

This pr does three things:
- adds an api for interacting with the SSL mTLS API (requires a new OAuth scope)
- uses the api for a new sub-command ('mtls-certificate') that allows uploading, listing, and deleting mTLS certificates
- supports new bindings (type: 'mtls_certificate') for worker uploads

NOTE: we are still waiting on the OAuth scope to be granted for Wrangler, so this shouldn't go just yet.

Associated docs issues/PR: https://github.com/cloudflare/cloudflare-docs/pull/7371

Author has included the following, where applicable:

- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [x] Checked for inclusion of relevant tests
- [x] Checked for inclusion of a relevant changeset
- [x] Checked for creation of associated docs updates
- [x] Manually pulled down the changes and spot-tested

Fixes # [insert issue number].

### CODE REVIEW

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (JacobMGEvans) [inline packages/wrangler/src/mtls-certificate/cli.ts:128]: confirmation path for `id` & `name` are hit but seems like the ternary isn't triggered for just the `id` path, the logger below that is triggered by the 2nd return is also not covered. Please provide a test scenario for this too.

**Review by JacobMGEvans — DISMISSED** [review]
Overall looking great, there are some negative pathing branch logic I would like to see in the tests and maybe get another pair of eyes on this, appreciate the API implementation! 😃

File: packages/wrangler/src/__tests__/publish.test.ts
Comment (IgorMinar) [inline packages/wrangler/src/__tests__/publish.test.ts:140]: Will this change require users to re-authenticate before they can use the `mtls-certificate` feature? I assume so since new permission is needed.

If so, will we gracefully handle this scenario or what will the experience be like for someone with an existing token that now tries to upload a certificate?

**Review by IgorMinar — COMMENTED** [review]
Thanks for changing the command name to a less ambiguous `mtls-certificate`, @matthewdavidrodgers.

File: packages/wrangler/src/__tests__/publish.test.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/__tests__/publish.test.ts:140]: Yes, unfortunately. even adding the permission to an existing scope would require re-authenticating as i'm told.

It would be simple enough to drop a try/catch around these methods and display a message like
```
Error: Unauthorized
You may need to log out and log back in
```

I could also detect when the user is using an OAuth token, check to see if it includes the required scope, and if not, prompt for a log out + back in

Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/publish.test.ts:140]: 🤔 I'll have to look into this more.

Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/publish.test.ts:140]: @IgorMinar by "central" I think you mean that we catch an authorisation error in the `cfetch()` calls and print a suitable message similar to what Matt suggests above?

**Comment by petebacondarwin** [comment]
We need an associated docs update at https://github.com/cloudflare/cloudflare-docs to go with this new feature.

File: packages/wrangler/src/__tests__/publish.test.ts
Comment (IgorMinar) [reply packages/wrangler/src/__tests__/publish.test.ts:140]: Yes. If we detect that the request has failed because the token is expired or doesn't have sufficient privileges, could we error gracefully with a actionable error message or even go as far as re-initiate the auth workflow?

Comment (IgorMinar) [reply packages/wrangler/src/__tests__/publish.test.ts:140]: What I meant by "central" was that this should be a Wrangler feature and not something that mTLS has to implement specifically for their command. Apologies about the ambiguity.

**Comment by matthewdavidrodgers** [comment]
Companion docs PR: https://github.com/cloudflare/cloudflare-docs/pull/7371
Just rebased to update the new oauth scope, we're renaming it before it's out in production to make it match naming conventions. The scope is still not fully released, so this should not merge yet.

**Comment by lrapoport-cf** [comment]
> Companion docs PR: [cloudflare/cloudflare-docs#7371](https://github.com/cloudflare/cloudflare-docs/pull/7371) Just rebased to update the new oauth scope, we're renaming it before it's out in production to make it match naming conventions. The scope is still not fully released, so this should not merge yet.

@matthewdavidrodgers thanks for providing the link to the companion docs PR! i added it to the PR summary comment at the top 👍

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:32]:
```suggestion
	yargs: CommonYargsArgv
```
(imported from `yarg-types.ts`)

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:33]:
```suggestion
) {
```

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:52]:
```suggestion
const uploadMTlsCertificateHandler = withConfig<StrictYargsOptionsToInterface<typeof uploadMTlsCertificateOptions>>(
```

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:98]: As above

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:111]: As above

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (penalosa) [inline packages/wrangler/src/mtls-certificate/cli.ts:158]: This should be an identity function `a => a`

**Review by penalosa — COMMENTED** [review]
A couple things related to tightening up the typing

File: packages/wrangler/src/__tests__/publish.test.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/publish.test.ts]: I will talk to Josh about how it kicks back an error, then we can look into a Core feature that can potentially do this "even go as far as re-initiate the auth workflow"

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/mtls-certificate/cli.ts]: rebased up and implemented these suggestions - thanks for the tip!

**Comment by matthewdavidrodgers** [comment]
Ok - the OAuth piece is finally finished. I have to touch up the docs pr, but everything's ready on this end after that. Is this blocked by the re-authentication issue, or can this go without having an answer to that piece yet?

**Comment by JacobMGEvans** [comment]
> Ok - the OAuth piece is finally finished. I have to touch up the docs pr, but everything's ready on this end after that. Is this blocked by the re-authentication issue, or can this go without having an answer to that piece yet?

No, it's not blocked on the re-auth, it has a ticket and is considered a DX improvement, which needs to be added to Wrangler Core.

**Comment by matthewdavidrodgers** [comment]
still chasing reviews from the docs side of things, but this is all good from my perspective

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/mtls-certificate/cli.ts:24]:
```suggestion
				"The path to a certificate file (.pem) containing a chain of certificates to upload",
```

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/mtls-certificate/cli.ts:29]:
```suggestion
			describe: "The path to a file containing the private key for your leaf certificate",
```

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/mtls-certificate/cli.ts:64]:
```suggestion
const listMTlsCertificatesHandler = withConfig(async ({ config }) => {
```

File: packages/wrangler/src/mtls-certificate/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/mtls-certificate/cli.ts:95]:
```suggestion
const deleteMTlsCertificateHandler = withConfig<
```

**Review by petebacondarwin — APPROVED** [review]
This is a sweet PR @matthewdavidrodgers - thanks for iterating through the feedback.
I have left a few more suggestions but nothing blocking from my side.

**Comment by petebacondarwin** [comment]
Last thing... I tried to upload a key and get an error from the CF API.
I generated a key pair using:

```sh
openssl genpkey -algorithm RSA -outform PEM -out private_key.pem -pkeyopt rsa_keygen_bits:2048
ssh-keygen -e -f private_key.pem > public_key.pem
```

Then tried uploading via:

```
npx wrangler mtls-certificate upload --cert public_key.pem --key private_key.pem --name test 
```

And I got the following error:

```
Uploading mTLS Certificate test...

✘ [ERROR] A request to the Cloudflare API (/accounts/3057da790214b3f8023628d382636e03/mtls_certificates) failed.

  Unable to parse certificate [code: 1408]
```

Is this expected? Did I do something wrong?

**Comment by matthewdavidrodgers** [comment]
@petebacondarwin checking the man page for ssh-keygen says that it will spin off a RFC4716 public key based on the private key. that's not what we want for the cert - we want a X509 client certificate (also in a PEM format), which encodes more data about an identity.

FWIW this is how i spin off client certs:
```
﻿#!/bin/bash

if [ $# -ne 1 ]; then
  echo 'must provide name for output'
  exit 1
fi

if [ -f "$1-cert.pem" ]; then
  echo "certs already exist; delete them or pick a new name"
  exit 1
fi

if [ ! -f "root.crt" ]; then
  echo "no root.crt file present"
  exit 1
fi

if [ ! -f "root.key" ]; then
  echo "no root.key file present"
  exit 1
fi

# generate key and csr
openssl req -newkey rsa:2048 -days 1825 -nodes -subj "/C=US/ST=California/L=Los Angeles/O=MDR/OU=mtlsclient/CN=mtlsclient" -keyout $1-key.pem -out $1-req.csr

# generate cert (will prompt for root.key passphrase)
openssl x509 -req -in $1-req.csr -CA root.crt -CAkey root.key -set_serial 01 -out $1-cert.pem

# clean up
rm $1-req.csr

echo ""
echo "upload this cert: npx wrangler mtls-certificate upload --cert $1-cert.pem --key $1-key.pem --name <name>"
```

you will need a root ca and key already generated.
you can also use cfssl, sometimes it's a bit easier (let me know if you want a hand getting some examples, this can be fickle)

**Comment by petebacondarwin** [comment]
Thanks for the info. As far as I am concerned Wrangler is doing the right thing, even if I generated the keys incorrectly. So I am happy to approve this.

File: packages/wrangler/src/api/mtls-certificate.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/api/mtls-certificate.ts:57]: Added short documentation to each function/exported interface, can expand if needed

**Comment by matthewdavidrodgers** [comment]
looks like the test failure is because of a missed timeout for unstable_dev in local mode? not related to this work, and i don't seem to be able to re-run the job

**Comment by lrapoport-cf** [comment]
@matthewdavidrodgers i just initiated a re-run of the failed jobs 👍

**Comment by matthewdavidrodgers** [comment]
more unstable_dev timeouts - seems unrelated

**Comment by matthewdavidrodgers** [comment]
docs PR is finally approved now - i can merge when needed. so ready to finally get the wrangler side merged!

**Comment by matthewdavidrodgers** [comment]
Docs PR merged - just rebased back over main, should be good to go from here if everything's green

**Comment by matthewdavidrodgers** [comment]
hmm... failure to pull workerd in both failures? not sure what to do here, hoping that's just flaky

**Pull request merged**
